man/man1/vmcp.1
.\" ############################################################################
.\" ### /usr/local/share/man/man1/vmcp.1.gz BEGIN
.\" ############################################################################
.\"
.\" Copyright 2017 WhiteWinterWolf (www.whitewinterwolf.com)
.\"
.\" This file is part of vmtools.
.\"
.\" vmtools is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" ############################################################################
.
.Dd June 3, 2017
.Dt VMCP 1
.Os vmtools
.
.
.Sh NAME
.
.Nm vmcp ,
.Nm vmfork
.Nd copy or fork virtual machines
.
.
.Sh SYNOPSIS
.
.Nm vmcp
.Op Fl hkqRrsvyz
.Op Fl M Ar module_name
.Op Fl m Ar module_name
.Op Fl o Ar setting Ns = Ns Ar value
.Op Fl a Oo Fl n Oc | Fl c | f
.Ar source ... directory
.Nm vmcp
.Op Fl hkqRrsvyz
.Op Fl M Ar module_name
.Op Fl m Ar module_name
.Op Fl o Ar setting Ns = Ns Ar value
.Op Fl a Oo Fl n Oc | Fl c | f
.Ar source destination
.Pp
.Nm vmfork
.Op Fl hkqRrsvyz
.Op Fl M Ar module_name
.Op Fl m Ar module_name
.Op Fl o Ar setting Ns = Ns Ar value
.Op Fl a Oo Fl n Oc | Fl c | f
.Ar source destination ...
.
.
.Sh DESCRIPTION
.
Create a copy of an currently existing virtual machine.
.Pp
Several copy modes are supported, in each mode each copied virtual machine has
a similar content than
.Ar source :
.Bl -tag -width "Autonomous copy:" -offset indent
.It Default copy :
The resulting virtual machines shares the same parent as its source.
This is the default operation mode for
.Nm vmcp .
.It Fork :
The resulting virtual machine is a child of
.Ar source .
This is the default operation mode for
.Nm vmfork .
.It Autonomous copy :
The resulting virtual machined does not depend on any parent VM.
.El
.Pp
The only differences between
.Ar source
and the resulting virtual machines are handled by the customizables
.Em clone
modules.
By default they only affect, if applicable, the virtual machine MAC address and
its storage path.
For more information on customizable modules, see
.Xr vmtools 7 .
.Pp
.Ar source
must be a valid virtual machine home directory (unless
.Fl r
has been provided) and the virtual machine must not be currently running.
.Pp
If only one
.Ar source
and one
.Ar destination
are provided, the resulting virtual machine home directory will be
.Ar destination
if it does not already exists, or a subdirectory of it if
.Ar destination
is an already existing directory.
.Pp
If more than two paths are provided:
.Bl -bullet -offset indent
.It
.Nm vmcp
expects the latest path to be a writable directory, a copy of each
.Ar source
virtual machines will be created below
.Ar directory.
.It
.Nm vmfork
expects the first path to be the
.Ar source
vitual machine, all subsequent paths are handled as
.Ar destination
paths to create as many copies of the
.Ar source
virtual machine.
.El
.
.
.Sh OPTIONS
.
.Ss Copy mode selection options
Each of these flags is valid for both
.Nm vmcp
and
.Nm vmfork ,
which means that it is possible to create a fork using
.Nm vmcp
by using the
.Fl f
flag and create default copies using
.Nm vmfork
by using the
.Fl c
flag.
This allows to take advantage of the different behavior of these commands in
presence of multiple
.Ar source
or
.Ar destination
arguments.
.Pp
The copy mode selection options are as follows:
.
.Bl -tag -width Ds
.It Fl a
Autonomous copy.
.br
The resulting virtual machine will not rely on any parent.
.
.It Fl c
Default copy.
.br
The resulting virtual machine will rely on the same parent as the source one.
If the source virtual machine has no parent, using either
.Fl a
or
.Fl c
produces the same result.
.
.It Fl f
Fork.
.br
The resulting virtual machine will use the source one as parent.
.El
.
.Ss Other options
.
The other options are as follows:
.
.Bl -tag -width Ds
.It Fl h
Show usage information summary then exit.
.
.It Fl k
Keep parent's unique properties.
.br
The
.Em clone
modules are still called to do the required processing to ensure the source
virtual machine integrity (like handling storage image files), however the
resulting virtual machines will share the same unique properties (like the MAC
address) than the source.
.
.It Fl M Ar module_name
Do not invoke the clone module
.Ar module_name .
.br
This is equivalent than removing
.Ar module_name
from the
.Cm cfg_modules_clone
setting (see
.Xr vmtools.conf 5 ) .
.br
.Ar module_name
may not exist and may not be present in
.Cm cfg_modules_clone .
.
.It Fl m Ar module_name
Invoke the clone module
.Ar module_name .
.br
This is equivalent than adding
.Ar module_name
to the
.Cm cfg_modules_clone
setting (see
.Xr vmtools.conf 5 ) .
.br
.Ar module_name
must exist, it may be already present in
.Cm cfg_modules_clone
in which case the module will be invoked only once.
.
.It Fl n
Do not modify the source virtual machine during an autonomous copy.
.br
Using this setting is generally not recommended unless the parent virtual
machine is located on some read-only storage where any lock attempt would fail.
.Pp
This setting can be used only in conjunction with
.Fl a
(autonomous copy).
.
.It Fl o Ar setting Ns = Ns Ar value
Affect
.Ar value
to
.Ar setting ,
overriding any previous value set in
.Xr vmtools 7
configuration or virtual machine settings.
Available settings are listed in the
.Xr vmtools.conf 5
file.
.
.It Fl q
Decrease verbosity.
Add several
.Fl q
options to decrease verbosity even more
.Fl ( qq
by default to get minimal output).
See
.Xr vmtools.conf 5
for more information about verbosity levels.
.Pp
See
.Fl v
to increase the verbosity level.
.
.It Fl R
Recurse over child virtual machines: each child of
.Ar source
is recursively copied as a new child of the destination virtual machine.
.br
Each of these new child virtual machines is created as a subdirectory of its
parent, independantly of the location of it matching source.
.Pp
See also
.Fl r
for directory-based recursion, both recursion modes can be associated.
.
.It Fl r
Recurse over directory content: each subdirectory of
.Ar source
containing a virtual machine home directory is copied below
.Ar destination
(including
.Ar source
itself if it is a virtual machine home), keeping relative paths intact.
.br
When this option is used,
.Ar source
only requires to be directory, it does not need to be virtual machine home.
.Pp
See also
.Fl R
for child-based recursion.
Both recursion modes can be associated, in such a case directory-based
recursion takes precedence to preserve the directory tree layout even if this
means that a child may not be a direct subdirectory of its parent.
.
.It Fl s
Enable Qemu snapshot (non-persistent) mode for the resulting virtual machines.
.br
The created virtual machines are volatile and do not store any change to their
storage backends.
Restarting such a virtual machine effectively rolls it back to a state similar
to the source virtual machine. This is useful to span virtual machines for
quick tests purpose from a clean master virtual machine.
.Pp
When using this option, when possible
.Nm vmcp
and
.Nm vmfork
do not create any disk image at all for the resulting virtual machines but
instead make them directly access the source virtual machine storage in
snapshot mode.
.Pp
This parameter has no effect on read-only storage.
.Pp
.Em Warning:
when manipulating sensitive data, ensure to always work on a copy instead of
the original file and/or unset the write bit on the backing file (see
.Xr chmod 1 ) .
.
.It Fl v
Increase verbosity.
Add several
.Fl v
options to increase verbosity even more
.Fl ( vvv
by default to get the most verbose output, including debugging messages).
See
.Xr vmtools.conf 5
for more information about verbosity levels.
.Pp
See
.Fl q
to decrease the verbosity level.
.
.It Fl y
Don't ask any question: accept all confirmation requests and automatically
select the default answer in any other situation.
.Pp
Use this option with
.Em great care :
no confirmation will be asked before deleting or overwriting any files!
.
.It Fl z
Compress copied or converted disk image files.
.br
This only affects QCow2 hard-disk images being copied or converted, data
written by a running guest is never compressed.
.Pp
Before creating a compressed copy, ensure from wihtin the guest that the
storage is correctly trimmed (on Linux systems, see
.Xr fstrim 8 )
as this shrinks the storage files to their minimum size.
.El
.
.
.Sh ENVIRONMENT
.
.Bl -tag -width Ds
.It Ev TMPDIR
Default location to store temporary files, by default
.Pa /tmp .
.
.It Ev XDG_CONFIG_HOME
Location of user's configuration files, by default
.Pa ~/.config .
.El
.
.
.Sh FILES
.
.Bl -tag -width Ds
.It Pa /usr/local/lib/vmtools
Libraries shared by the vmtools project utilities.
.It Pa /usr/local/share/vmtools/modules/clone
Modules handling the virtual machine copying process.
.It Pa /usr/local/share/vmtools/vmtools.conf
Virtual machine default settings, see
.Xr vmtools.conf 5 .
.El
.Pp
Moreover, the content of
.Pa /usr/local/share/vmtools
can be overridden in the following locations (in the order of precedence):
.Bl -tag -width Ds
.It ~/.config/vmtools
User overrides (if
.Cm cfg_include_userhome
is set to
.Dq yes ) .
.It /etc/vmtools
System-wide overrides.
.El
.
.
.Sh EXIT STATUS
.
.Bl -tag -compact -width Ds
.It 0
The operation ended successfully.
.It 1
An error occurred, incomplete tasks have been reverted.
.It 2
Wrong usage, nothing has been done.
.El
.
.
.Sh EXAMPLES
.
.Ss Nm vmfork
.
.Bl -inset
.It Example 1:
Fork a virtual machine:
.Pp
.Dl Ic vmfork ./fooOS-base ./fooOS-20170515
.Pp
.It Example 2:
Create a snapshot fork of a virtual machine:
.Pp
.Dl Ic vmfork -s ./foo ./foo-tests
.Pp
.It Example 3:
Create several forks of one virtual machine (compare it to the
.Nm vmcp
example 2, the relationship here is one-to-many):
.Pp
.Dl Ic vmfork ./foo ./fork1 ./fork2 ./fork3
.Pp
.El
.
.Ss Nm vmcp
.Bl -inset
.It Example 1:
Copy several virtual machines to a directory:
.Pp
.Dl Ic vmcp ./foo ./bar ~/archives
.Pp
.It Example 2:
Create a fork of all virtual machines located directly below a
directory (compare it to the
.Nm vmfork
example 3, the relationship here is many-to-one):
.Pp
.Dl Ic vmcp -f ~/archived_vms/* ~/work
.Pp
.It Example 4:
Recursively copy all the virtual machines stored in a directory:
.Pp
.Dl Ic vmcp -r ~/Documents/VMs ~/backup
.Pp
.El
.
.
.Sh SEE ALSO
.
.Xr vmmerge 1 ,
.Xr vmmv 1 ,
.Xr vmrm 1 ,
.Xr vmtools 7
.
.
.Sh REPORTING BUGS
.
Please send bug reports to the
.Lk http://github.com/WhiteWinterWolf/vmtools/issues "vmtools issues page" .
.
.\" ############################################################################
.\" ### /usr/local/share/man/man1/vmcp.1.gz END
.\" ############################################################################


lib/vmcp.inc.sh
################################################################################
### /usr/local/lib/vmtools/vmcp.inc.sh BEGIN
################################################################################
#
# Copyright 2017 WhiteWinterWolf (www.whitewinterwolf.com)
#
# This file is part of vmtools.
#
# vmtools is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# ------------------------------------------------------------------------------
#
# This library handles the copy of a virtual machine from one location to
# another.
#
# Several copy modes are supported, in each mode the resulting VM will have a
# similar content than its source:
#   - Default copy: the resulting VM shares the same parent as its source.
#   - Fork: the resulting VM is a child of the source VM.
#   - Autonomous copy: the resulting VM does not depend on any parent VM.
#
# The only differences between the the source and destination VM are handled
# by the customizables "clone" modules (by default it affects storage image
# path and network MAC address).
#
# The copy may be recursive, two kinds of recursion are supported:
#   - Child-based recursion: all source's childs will be copied as
#     destination's childs.
#   - Directory-based recursion: all VM located below the source path (which
#     may in this case not necessarily be a VM home directory) will be copied
#     below the destination path.
#
# These two recursion types can be combined to execute a child-based recursive
# copy of all VM located below a path.
#
# See `vmcp_copy_single()' for implementation details on the various copy types.
#
# Public functions:
#   vmcp_copy source target
#         Copy the virtual machine `source' to `target'.
#   vmcp_parseargs_args arg...
#         Parse command-line arguments, sets `$vmcp_*' global variables.
#
################################################################################

################################################################################
# Global variables
################################################################################

# The settings below should be set using the `vmcp_parseargs_args()' function.

# Name of action to perform: 'copy', 'autonomous copy' or 'fork'.
# This variable is also used by the clone modules to determine what to do.
vmcp_action=''

# If set to 'yes', parent's unique properties like the MAC address will be
# inherited by the child VM.
vmcp_clone_inherit='no'

# If "yes", enable recursion over virtual machine childs.
vmcp_recurse_childs='no'

# If "yes", enable recursion over given path subdirectories.
vmcp_recurse_directories='no'

# If "yes", use snapshot for all writable storage backends.
vmcp_snapshot='no'

# If "yes", the source VM is considered read-only.
vmcp_readonly='no'


################################################################################
# Functions
################################################################################

###
# vmcp_copy source target
#
# Copy the virtual machine `source' to `target'.
#
# If `target' is an already existing directory, `source' will be created as a
# subdirectory of `target', otherwise `target' will directly contain the copied
# virtual machine.
#
# SC2030: This function executes itself in a subshell.
# shellcheck disable=SC2030
#
vmcp_copy() (
	local 'action_old' 'dst' 'lock_opt' 'recurse_done'
	local 'recurse_dst_root' 'recurse_src_root' 'src' 'src_parent_path'
	[ "${1-}" = '--' ] && shift
	cleanup_reset

	# Set the source directory.
	src=${1:?"ERROR (BUG): vmcp_copy: Missing parameter."}
	src=$( realpath -- "$src" ) || return 1

	# Set the destination directory.
	dst=${2:?"ERROR (BUG): vmcp_copy: Missing parameter."}
	if [ -d "$dst" ]
	then
		dst=$( storage_createpath -d -t "$dst" -- "$src" ) || return 1
	else
		dst=$( storage_createpath -d -t "$dst" -- '' ) || return 1
	fi
	homedir_init -- "$dst" || return 1
	dst=$vm_home

	# Lock the source directory.
	lock_opt=''
	[ "$vmcp_recurse_childs" = 'yes' ] && lock_opt="${lock_opt}c"
	[ "$vmcp_recurse_directories" = 'yes' ] && lock_opt="${lock_opt}d"
	if [ "$vmcp_readonly" = 'yes' ]
	then
		lock_acquire ${lock_opt:+"-${lock_opt}"} -n -- "$src" || return 1
	else
		lock_acquire ${lock_opt:+"-${lock_opt}"} -- "$src" || return 1
		src_parent_path=$( parent_get_nearest -- "$src" ) || return 1
		if [ -n "$src_parent_path" ]
		then
			lock_acquire -- "$src_parent_path" || return 1
		fi
	fi

	# Copy the VM.
	action_old=$vmcp_action
	recurse_done=''
	recurse_dst_root=$dst
	recurse_src_root=$src
	vmcp_copy_recurse -- 1 "$src" "$dst" || return 1

	# Unlock both the source and destination directories.
	if [ "$vmcp_readonly" != 'yes' ]
	then
		if [ -n "$src_parent_path" ]
		then
			lock_release -- "$src_parent_path" || return 1 1
		fi
		lock_release ${lock_opt:+"-${lock_opt}"} -- "$src" || return 1
	fi
	lock_release -d -- "$dst" || return 1
	cleanup_end
)

###
# vmcp_copy_recurse level source destination
#
# This is a helper and is not designed to be called outside of this library.
#
# This function implements recursion, the copy itself is delegated to
# `vmcp_copy_single()'.
#
# SC2031: This function is invoked by `vmcp_cp()', it has access to its
# variables.
# shellcheck disable=SC2031
#
vmcp_copy_recurse() {
	local 'dst' 'level' 'src' 'src_child' 'src_childs_list' 'src_parent_path'
	local 'src_subdir' 'src_subdirs_list' 'vm_home'
	[ "${1-}" = '--' ] && shift
	level=${1:?"ERROR (BUG): vmcp_copy_recurse: Missing parameter."}
	src=${2:?"ERROR (BUG): vmcp_copy_recurse: Missing parameter."}
	dst=${3:?"ERROR (BUG): vmcp_copy_recurse: Missing parameter."}

	level=$(( level + 1 ))
	if [ "$level" -ge "${cfg_limit_nesting:?}" ]
	then
		echo "ERROR: A loop has been detected during the copy process." >&2
		echo "Current source: '${src}'." >&2
		return 1
	fi

	if [ ! -d "$src" ]
	then
		echo "ERROR: ${src}: The source path is not a valid directory." >&2
		return 1
	elif test "$vmcp_readonly" != 'yes' && ! lock_check -- "$src"
	then
		 echo "ERROR: ${src}: The source path is not locked." >&2
		 return 1
	fi

	if ! printf '%s' "$recurse_done" | grep -Fqx "$src"
	then
		str_list_add 'recurse_done' "$src" || return 1

		# Copy directories and VMs.

		if [ "$vmcp_recurse_directories" = 'yes' \
		-a ! -e "${src}/${cfg_file_vmsettings:?}" ]
		then
			# Recursing over a directory which does not contain a VM:
			# Simply create an empty directory.
			dst="${dst}/${src##*/}"
			if [ ! -e "$dst" ]
			then
				mkdir -- "$dst" || return 1
			fi
			lock_acquire -- "$dst" || return 1
		else
			# If we are recursively copying a tree, ensure we copy the parent
			# first if it is located at a lower level to avoid dependency
			# issues.
			if [ "$vmcp_recurse_directories" = 'yes' ]
			then
				src_parent_path="$( parent_get_nearest -- "$src" )" || return 1
				if [ -n "$src_parent_path" ]
				then
					dst_parent_path=$( vmmv_mvpath -- "${recurse_src_root:?}" \
						"${recurse_dst_root:?}" "$src_parent_path" ) || return 1

					if test "$dst_parent_path" != "$src_parent_path" && \
						! printf '%s' "$recurse_done" \
						| grep -Fqx "$src_parent_path"
					then
						vmcp_action='recursive copy'
						vmcp_copy_recurse -- "$level" "$src_parent_path" \
							"$dst_parent_path" || return 1
					fi
				fi
			fi

			# Check `$dst' value.
			if [ "$src" = "${recurse_src_root:?}" ]
			then
				# Initial round, the destination should have been locked by
				# `vmcp_copy()'.
				lock_check -e -- "$dst" || return 1
				vm_home=$dst
			else
				# Following rounds: destination is locked by `homedir_init()'.
				dst=$( storage_createpath -h -t "$dst" -- '' "$src" ) \
					|| return 1
				homedir_init -- "$dst" || return 1
				# `$vm_home' contains the canonical path to `$dst'.
				dst=$vm_home
			fi

			vmcp_copy_single -- "$src" "$dst" || return 1
		fi

		# Handle recursion.

		# Process drectory-based recursion first as it must respect source
		# directories structure.
		if [ "$vmcp_recurse_directories" = 'yes' ]
		then
			src_subdirs_list=$( find "$src" -mindepth 1 -maxdepth 1 \
				\! -path "*${newline:?}*" -type d ) || return 1
			for src_subdir in $src_subdirs_list
			do
				vmcp_action='recursive copy'
				vmcp_copy_recurse -- "$level" "$src_subdir" "$dst" || return 1
			done
		fi

		# Process child-based recursion to complete directory-based
		# generated directory tree with external VMs.
		if [ "$vmcp_recurse_childs" = 'yes' \
			-a -e "${src}/${cfg_file_vmsettings:?}" ]
		then
			src_childs_list=$( childs_get_list -- "$src" ) || return 1
			for src_child in $src_childs_list
			do
				vmcp_action='recursive copy'
				vmcp_copy_recurse -- "$level" "$src_child" "$dst" || return 1
			done
		fi
	fi
}

###
# vmcp_copy_single source destination
#
# This is a helper for `vmcp_copy()' and is not designed to be called from
# outside of this library.
#
# This function implements the actual copy of the VM and the invocation of the
# "clone" modules through `vmcp_invoke_modules()'.
#
# SC2031: This function is invoked by `vmcp_cp()', it has access to its
# variables.
# shellcheck disable=SC2031
#
vmcp_copy_single() {
	local 'comment' 'dst' 'dst_parent_path' 'src' 'src_parent_path' 'vm_home'
	[ "${1-}" = '--' ] && shift
	src=${1:?"ERROR (BUG): vmcp_copy_single: Missing parameter."}
	dst=${2:?"ERROR (BUG): vmcp_copy_single: Missing parameter."}

	dst_parent_path=''
	src_parent_path=$( parent_get_nearest -- "$src" ) || return 1
	if [ "${vmcp_action:?}" = 'recursive copy' ]
	then
		if [ -n "$src_parent_path" ]
		then
			dst_parent_path=$( vmmv_mvpath -- "${recurse_src_root:?}" \
				"${recurse_dst_root:?}" "$src_parent_path" ) || return 1
		fi
		if [ "$dst_parent_path" = "$src_parent_path" ]
		then
			# The source's parent is located outside of the recursion tree:
			# We need to apply the initial action.
			vmcp_action=${action_old:?}
		fi
	fi

	case "${vmcp_action:?}" in
		'copy')
			cli_trace 3 "${src}: Copy to '${dst}'."
			settings_import -s -- "$src" || return 1
			comment="Copied from '${vm_name}':${newline}${src}"
			vm_home=$dst
			vm_name=${dst##*/}

			parent_clear
			if [ -n "$src_parent_path" ]
			then
				parent_add -- "$src_parent_path" || return 1
			fi

			vmcp_invoke_modules -- "$src" || return 1
			settings_save "$comment" "$dst" || return 1

			if [ "$vmcp_readonly" != 'yes' ]
			then
				if [ -n "$src_parent_path" ]
				then
					childs_add -- "$src_parent_path" "$dst" || return 1
					childs_save -- "$src_parent_path" || return 1
				fi
			fi
			;;

		'fork')
			cli_trace 3 "${src}: Fork to '${dst}'."
			# Do not use `settings_import()' as forked VM must contain only
			# modified settings and inherit the rest from their ancestors.
			settings_loadvm -- "$src" || return 1
			comment="Forked from '${vm_name}':${newline}${src}"
			vm_home=$dst
			vm_name=${dst##*/}

			parent_clear
			parent_add -- "$src" || return 1

			vmcp_invoke_modules -- "$src" || return 1
			settings_save "$comment" "$dst" || return 1

			childs_add -- "$src" "$dst" || return 1
			childs_save -- "$src" || return 1
			;;

		'recursive copy')
			settings_import -s -- "$src" || return 1
			comment="Recursive copy of '${vm_name}':${newline}${src}"
			vm_home=$dst
			vm_name=${dst##*/}

			# Update parent link
			parent_clear
			if [ -n "$src_parent_path" ]
			then
				if [ "$vmcp_recurse_directories" = 'yes' ]
				then
					cli_trace 3 "${src}: Directory-recursive copy to" \
						"'${dst}'."
					# Directory-based recursion directly transposes the
					# source directory structure below the recursion root.
					parent_add -- "${dst_parent_path:?}" || return 1
				elif [ "$vmcp_recurse_childs" = 'yes' ]
				then
					cli_trace 3 "${src}: Child-recursive copy to '${dst}'."
					# Child-based recursion creates a new directory
					# structure where VM childs are direct subfolders.
					dst_parent_path=${dst%/*}
					parent_add -- "${dst_parent_path:?}" || return 1
				else
					echo "ERROR (BUG): vmcp_copy: Selected action is" \
						"'recursive copy' but neither directory nor child" \
						"recursion is enabled." >&2
					return 1
				fi
			fi

			vmcp_invoke_modules -- "$src" || return 1
			settings_save "$comment" "$dst" || return 1

			if [ -n "$dst_parent_path" ]
			then
				childs_add -- "$dst_parent_path" "$dst" || return 1
				childs_save -- "$dst_parent_path" || return 1
			fi
			;;

		'autonomous copy')
			cli_trace 3 "${src}: Autonomous copy to '${dst}'."
			settings_import -- "$src" || return 1
			comment="Autonomous copy of '${vm_name}':${newline}${src}"
			vm_home=$dst
			vm_name=${dst##*/}

			parent_clear

			vmcp_invoke_modules -- "$src" || return 1
			settings_save "$comment" "$dst" || return 1
			;;

		*)
			echo "ERROR (BUG): vmcp_copy: Invalid value for \$vmcp_action:"\
				"'${vmcp_action}'." >&2
			return 1
			;;
	esac
}

###
# vmcp_invoke_modules src_vm
#
# This function is a helper for `vmcp_copy()' and is not designed to be called
# directly.
#
# Invoke the clone modules listed in `$cfg_modules_clone'.
#
vmcp_invoke_modules() {
	local 'mods_list' 'module' 'src'
	[ "${1-}" = '--' ] && shift
	src=${1:?"ERROR (BUG): vmcp_invoke_module: Missing parameter."}

	mods_list=$( str_explode -- "$cfg_modules_clone" ) || return 1
	for module in $mods_list
	do
		cli_trace 3 "${vm_home}: cloning VM settings: ${module}."
		include_module -- "clone/${module}.inc.sh" || return 1
		mod_clone "$src" || return 1
	done
}

###
# vmcp_parseargs_args arg...
#
# Parse command-line arguments and sets `$vmcp_*' global variables accordingly.
#
# Sets the global variable OPTIND to allow shifting the parameters in the
# calling context (see http://pubs.opengroup.org/onlinepubs/9699919799/utilities/getopts.html).
#
# The default value for `$vmcp_action' must be explicitely set by the caller
# *after* calling this function if it was not set by user parameters:
#     vmcp_action=${vmcp_action:-'copy'}
#
vmcp_parseargs() {
	# `$OPTIND' must not be available to the caller, do not set it local!
	local 'OPTARG'

	OPTIND=1
	while getopts "acfhkM:m:no:qRrsvyz" param
	do
		case "$param" in

			'a') # Produce an autonomous VM.
				if [ -n "$vmcp_action" ]
				then
					echo "ERROR: You cannot mix '-a', '-c' and '-f' flags." \
						>&2
					return 2
				fi
				vmcp_action='autonomous copy'
				;;

			'c') # Copy the source VM.
				if [ -n "$vmcp_action" ]
				then
					echo "ERROR: You cannot mix '-a', '-c' and '-f' flags." \
						>&2
					return 2
				fi
				vmcp_action='copy'
				;;

			'f') # Fork the source VM.
				if [ -n "$vmcp_action" ]
				then
					echo "ERROR: You cannot mix '-c', '-f' and '-s' flags." \
						>&2
					return 2
				fi
				vmcp_action='fork'
				;;

			'h') # Show usage information.
				printf '%s\n' "$usage"
				exit 0
				;;

			'k') # Keep parent's unique properties
				vmcp_clone_inherit='yes'
				;;

			'M') # Disable a clone module.
				case " ${cfg_modules_clone} " in
					*" ${OPTARG} "*)
						cfg_modules_clone=$( printf '%s' "$cfg_modules_clone" \
							| sed "s/$( str_escape_sed -- "$OPTARG" )//g" )
						;;
					*)
						cli_trace 4 "Duplicate module '${OPTARG}' already" \
							"disabled."
						;;
				esac
				;;

			'm') # Enable a supplementary clone module.
				case " ${cfg_modules_clone} " in
					*" ${OPTARG} "*)
						cli_trace 4 "Clone module '${OPTARG}' already enabled."
						;;
					*)
						cfg_modules_clone="$cfg_modules_clone $( \
							str_escape -- "$OPTARG" )"
						;;
				esac
				;;

			'n') # Do not modify the source VM during an autonomous copy.
				vmcp_readonly='yes'
				;;

			'o') # Override a setting.
				value=$( str_unescape -- "${OPTARG#*=}" ) || return 2
				settings_override "${OPTARG%%=*}" "$value" || return 2
				;;

			'q') # Decrease verbosity.
				if [ "${cfg_ui_verbosity:?}" -gt 0 ]
				then
					cfg_ui_verbosity=$(( cfg_ui_verbosity - 1 )) || return 1
				fi
				;;

			'R') # Recurse over VM childs.
				vmcp_recurse_childs='yes'
				;;

			'r') # Recurse over subdirectories.
				vmcp_recurse_directories='yes'
				;;

			's') # Snapshot mode.
				vmcp_snapshot='yes'
				;;

			'v') # Increase verbosity.
				cfg_ui_verbosity=$(( ${cfg_ui_verbosity:?} + 1 )) || return 1
				;;

			'y') # Never ask any confirmation, always assume `y' as answer.
				settings_override 'cfg_ui_assumeyes' 'yes'
				;;

			'z') # Compress copied or converted images.
				settings_set 'vm_qemu_compress' 'yes'
				;;

			*)
				printf 'Unexpected argument: %s\n' "$1" >&2
				return 2
				;;
		esac
	done
	shift $(( OPTIND - 1 ))

	settings_set 'cfg_ui_verbosity' "${cfg_ui_verbosity:?}"
	if [ "${cfg_ui_verbosity:?}" -ge 5 ]
	then
		set -x
	fi

	if [ "$vmcp_readonly" = 'yes' -a "$vmcp_action" != 'autonomous copy' ]
	then
		echo "ERROR: '-n' can only used with '-a' (autonomous copy mode)." >&2
		return 2
	fi

	if [ $# -lt 2 ]
	then
		echo "ERROR: You must provide at least one source and one target" \
			"path." >&2
		return 2
	fi
}

################################################################################
### /usr/local/lib/vmtools/vmcp.inc.sh END
################################################################################
